First, we need to read the data. I'll load the HEF and header data into pandas dataframes.

In [1]:
import os
import h5py
import numpy as np
import pandas as pd
from ntplib import system_to_ntp_time

pd.set_option('display.notebook_repr_html', True)

NTP_DIFF = system_to_ntp_time(0)

hef_file = os.path.join('data', 'horizontal_electric_field.nc')
header_file = os.path.join('data', 'hpies_data_header.nc')

def netcdf_to_pandas(path):
    h5file = h5py.File(path)
    times = h5file['time'].value
    times = times - NTP_DIFF
    times = pd.to_datetime(times, unit='s')
    data = {}
    for group in h5file:
        if group == 'time':
            continue
        for var in h5file[group]:
            if 'timestamp' in var:
                continue
            data[var] = h5file[group][var].value
    return pd.DataFrame(data, index=times)
    
hef_df = netcdf_to_pandas(hef_file)
header_df = netcdf_to_pandas(header_file)

Now, we can check the contents.

In [2]:
len(hef_df), len(header_df)

(12009, 252)

We have lots of HEF data and header data at around 1/50th the data rate. Let's re-index the header data to the same time series as the HEF data, forward filling the missing data points (using the last seen value).

In [3]:
header_df = header_df.reindex(index=hef_df.index, method='ffill')
len(header_df)

12009

Now the time series for both datasets match, let's join the two dataframes:

In [4]:
joined = hef_df.join(header_df, rsuffix='_header')
len(joined)

12009

Now joined contains the combination of the HEF data and header data. We now have a hpies_hcno value for all data points in the HEF data:

In [5]:
joined[hef_df.columns.values.tolist() + ['hpies_hcno']]

hpies_data_valid  hpies_e1a  hpies_e1b  \
2015-04-29 14:21:08.658715                    1     396607     396698   
2015-04-29 14:21:08.708307                    1     397149     397297   
2015-04-29 14:21:08.766100                    1     397208     397333   
2015-04-29 14:21:08.825862                    1     397201     397312   
2015-04-29 14:21:08.887132                    1     397217     397346   
2015-04-29 14:21:08.947684                    1     397223     397290   
2015-04-29 14:21:09.009053                    1     397214     397339   
2015-04-29 14:21:09.069960                    1     397198     397299   
2015-04-29 14:21:09.131373999                 1     397185     397311   
2015-04-29 14:21:09.191834                    1     397210     397304   
2015-04-29 14:21:09.253042                    1     397206     397291   
2015-04-29 14:21:09.318136                    1     397219     397312   
2015-04-29 14:21:09.379428                    1     397194     397325   
2015-04-29 14:21:09.440475                    1     397200     397314   
2015-04-29 14:21:09.505502                    1     397189     397298   
2015-04-29 14:21:09.566555                    1     397221     397293   
2015-04-29 14:21:09.627611                    1     397190     397335   
2015-04-29 14:21:09.692813                    1     397209     397313   
2015-04-29 14:21:09.753664                    1     397195     397293   
2015-04-29 14:21:09.818702                    1     397236     397315   
2015-04-29 14:21:09.880041                    1     397248     397322   
2015-04-29 14:21:09.940811                    1     397237     397330   
2015-04-29 14:21:10.002160                    1     397197     397314   
2015-04-29 14:21:10.066776                    1     397170     397302   
2015-04-29 14:21:10.127924                    1     397193     397324   
2015-04-29 14:21:10.188960                    1     397206     397273   
2015-04-29 14:21:10.253843                    1     397245     397328   
2015-04-29 14:21:10.315128                    1     397221     397308   
2015-04-29 14:21:10.375850                    1     397193     397336   
2015-04-29 14:21:10.441085                    1     397201     397300   
...                                         ...        ...        ...   
2015-04-29 19:23:13.564325                    1     397224     397355   
2015-04-29 19:23:13.629448                    1     397256     397360   
2015-04-29 19:23:13.690330                    1     397229     397331   
2015-04-29 19:23:13.751659                    1     397191     397358   
2015-04-29 19:23:13.816609                    1     397185     397340   
2015-04-29 19:23:13.877603                    1     397202     397416   
2015-04-29 19:23:13.938618                    1     397243     397378   
2015-04-29 19:23:14.003849                    1     397207     397384   
2015-04-29 19:23:14.064804                    1     397199     397405   
2015-04-29 19:23:14.126094                    1     397253     397362   
2015-04-29 19:23:14.186773                    1     397263     397345   
2015-04-29 19:23:14.251910                    1     397194     397351   
2015-04-29 19:23:14.313589                    1     397204     397327   
2015-04-29 19:23:14.374279                    1     397218     397353   
2015-04-29 19:23:14.439219                    1     397213     397334   
2015-04-29 19:23:14.501193                    1     397203     397325   
2015-04-29 19:23:14.561806                    1     397231     397365   
2015-04-29 19:23:14.622534                    1     397199     397354   
2015-04-29 19:23:14.687579                    1     397202     397327   
2015-04-29 19:23:14.752342                    1     397183     397368   
2015-04-29 19:23:14.809747                    1     397209     397359   
2015-04-29 19:23:14.874367                    1     397223     397358   
2015-04-29 19:23:14.935499                    1     397233     397358   
2015-04-29 19:23

Let's compute the data rate for the HEF data:

In [6]:
data_rate = 1 / np.median(np.diff(joined.index.values).astype('f64') / 1e9)
'data rate: %.2f Hz' % data_rate

'data rate: 16.32 Hz'

Now, let's compute the storage requirements for the lifetime of the project if we are to add just one single 32-bit integer to each HEF particle:

In [7]:
one_day = data_rate * 60 * 60 * 24
one_year = one_day * 365
project_lifetime = one_year * 25
int_storage = project_lifetime * 4 / 1e9
'Expected storage cost for adding one int to each particle: %.2f GB' % int_storage 

'Expected storage cost for adding one int to each particle: 51.48 GB'

So, by storing the header data separately and putting the onus on the user to combine the datasets we save over 51 GB (just for the hcno data alone.)

Now, let's see if we can figure out what went wrong to produce the occasional jump in eindex.

In [8]:
import ntplib
import struct
import time
import os

message_types = ( 'Data From Instrument',
                  'Data From Driver',
                  'Port Agent Command',
                  'Port Agent Status',
                  'Port Agent Fault',
                  'Instrument Command',
                  'Heartbeat',
                  'Pickled Data From Instrument',
                  'Pickled Data From Driver' )

HEADER_FORMAT = '>3sBHHII'
SYNC_BYTES = '\xa3\x9d\x7a'
raw_data_fh = open(os.path.join('data', 'hpies-port_agent.20150429.data'))
raw_data = raw_data_fh.read()

In [9]:
def ntp_time_to_ascii(tstamp):
    return time.asctime(time.gmtime(ntplib.ntp_to_system_time(tstamp)))

def parse_header(data):
    sync, message_type, packet_size, checksum, time_upper, time_lower = struct.unpack_from(HEADER_FORMAT, data)
    timestamp = time_upper + float(time_lower) / 2**32
    return message_type, packet_size, checksum, timestamp

Data from the instrument is message type 1

In [10]:
HEADER_SIZE = struct.calcsize(HEADER_FORMAT)
packets = []
while raw_data:
    start = raw_data.find(SYNC_BYTES)
    if start == -1:
        break
    
    message_type, packet_size, checksum, timestamp = parse_header(raw_data[start:])
    packet = raw_data[start:start+packet_size]
    raw_data = raw_data[start+packet_size:]
    
    if message_type == 1:
        packets.append((timestamp, packet))

In [11]:
import sys
NEWLINE = '\r\n'
records = []
buf = ''

for index, packet in enumerate(packets):
    timestamp, data = packet
    if index == 0:
        tstamp = timestamp
        

    buf += data[HEADER_SIZE:]
    if NEWLINE in buf:
        n_index = buf.index(NEWLINE)
        records.append((tstamp, buf[:n_index]))
        buf = buf[n_index+len(NEWLINE):]
        tstamp = timestamp

In [12]:
HEF_records = [r for r in records if r[1].startswith('#3__DE')]
times = np.array([r[0] for r in HEF_records]) - NTP_DIFF
times = pd.to_datetime(times, unit='s')
raw_hef_df = pd.DataFrame([r[1] for r in HEF_records], index=times)
len(raw_hef_df) - len(joined)

24

There are 24 more records in the source data than the parsed data. This matches the number of jumps:

In [13]:
eindex = joined.hpies_eindex.values
jumps = np.diff(eindex) == 20
len(eindex[jumps])

24

so, what's up with these records?

Let's build a new jumps indexed to the raw data by inserting another item after each jump...

In [14]:
new_jumps = []
for jump in jumps:
    new_jumps.append(jump)
    if jump:
        new_jumps.append(True)
        
new_jumps.append(False)
        
pd.set_option('max_colwidth',100)
jump_df = raw_hef_df[new_jumps]
jump_df

0
2015-04-29 15:33:09.045122        #3__DE 59 396663 397170 397282 397417 396834 397734*2a72
2015-04-29 15:33:09.106457      #3__DE 69 396672 397189 397303 397438 396820 397744\r*23f7
2015-04-29 15:51:09.297572        #3__DE 89 396670 397219 397320 397410 396820 397780*2f49
2015-04-29 15:51:09.363066      #3__DE 99 396608 397176 397380 397444 396801 397774\r*f8f6
2015-04-29 15:59:08.987504        #3__DE 39 396702 397227 397332 397406 396827 397772*0f21
2015-04-29 15:59:09.123707        #3__DE 49 396688 397234 397354 397451 396786 397788*8ebd
2015-04-29 16:05:09.367980        #3__DE 89 396630 397215 397357 397455 396827 397783*de97
2015-04-29 16:05:09.500727999     #3__DE 99 396668 397191 397333 397413 396832 397757*c653
2015-04-29 16:15:12.684828       #3__DE 629 396631 397200 397350 397442 396814 397777*f296
2015-04-29 16:15:12.870204       #3__DE 639 396700 397181 397355 397443 396812 397769*ba65
2015-04-29 16:15:13.975050       #3__DE 829 396666 397248 397361 397433 396817 397777*d887
2015-04-29 16:15:14.035774       #3__DE 839 396695 397180 397384 397441 396799 397778*3d8d
2015-04-29 16:19:10.065224       #3__DE 209 396682 397221 397387 397418 396822 397778*98d9
2015-04-29 16:19:10.132692       #3__DE 219 396693 397250 397330 397445 396807 397762*820a
2015-04-29 16:19:12.551835       #3__DE 599 396688 397199 397355 397450 396821 397742*8655
2015-04-29 16:19:12.555627       #3__DE 609 396713 397236 397375 397449 396834 397781*e8f0
2015-04-29 16:29:14.304653       #3__DE 889 396665 397201 397372 397452 396819 397772*ec40
2015-04-29 16:29:14.436729       #3__DE 899 396677 397319 397355 397451 396799 397770*e6b6
2015-04-29 16:37:13.759633       #3__DE 779 396601 397222 397323 397441 396826 397786*e95c
2015-04-29 16:37:13.830661       #3__DE 789 396698 397210 397333 397431 396833 397764*3724
2015-04-29 16:45:10.405998       #3__DE 259 396683 397220 397320 397421 396790 397779*f27f
2015-04-29 16:45:10.550904       #3__DE 269 396623 397219 397340 397446 396834 397793*1f17
2015-04-29 16:53:09.252755999     #3__DE 49 396662 397230 397321 397439 396823 397771*0356
2015-04-29 16:53:09.256195        #3__DE 59 396660 397197 397310 397424 396799 397783*23f7
2015-04-29 16:59:09.394574        #3__DE 89 396702 397214 397370 397436 396798 397783*17bc
2015-04-29 16:59:09.455009      #3__DE 99 396658 397211 397384 397439 396827 397790\r*e0e9
2015-04-29 17:17:11.879393       #3__DE 489 396715 397154 397360 397457 396801 397803*3c41
2015-04-29 17:17:12.007452       #3__DE 499 396680 397249 397359 397413 396781 397762*4c13
2015-04-29 17:19:11.738710       #3__DE 459 396679 397205 397357 397471 396804 397773*be95
2015-04-29 17:19:11.742570       #3__DE 469 396657 397216 397342 397431 396826 397782*6d9b
2015-04-29 17:39:14.171886       #3__DE 849 396676 397227 397358 397412 396808 397801*2c1e
2015-04-29 17:39:14.310836       #3__DE 859 396650 397186 397338 397407 396816 397775*2f69
2015-04-29 17:43:09.832230       #3__DE 139 396658 397208 397362 397445 396822 397793*4db8
2015-04-29 17:43:09.836858       #3__DE 149 396724 397187 397360 397480 396823 397747*d9f8
2015-04-29 18:05:13.977414       #3__DE 799 396695 397220 397360 397420 396820 397768*1748
2015-04-29 18:05:14.109137       #3__DE 809 396676 397230 397354 397466 396796 397771*075e
2015-04-29 18:39:14.808507       #3__DE 929 396691 397223 397366 397394 396807 397775*aaa6
2015-04-29 18:39:14.813237       #3__DE 939 396696 397223 397374 397486 396815 397770*db0b
2015-04-29 18:53:09.611510        #3__DE 99 396653 397171 397364 397479 396850 397753*97a3
2015-04-29 18:53:09.668377     #3__DE 109 396706 397248 397309 397469 396847 397777\r*57e4
2015-04-29 18:53:11.228877       #3__DE 349 396655 397220 397361 397479 396855 397801*9470
2015-04-29 18:53:11.232502       #3__DE 359 396670 397208 397342 397444 396824 397780*6bb6
2015-04-29 19:03:09.612759        #3__DE 99 396655 397203 397374 397444 396817 397784*7cf0
2015-04-29 19:03:09.670427     #3__DE 109 396635 397174 397390 397503 396827 397773\r

In [15]:
joined.hpies_eindex[jumps]

2015-04-29 15:33:09.045122        59
2015-04-29 15:51:09.302055        89
2015-04-29 15:59:08.991716        39
2015-04-29 16:05:09.367980        89
2015-04-29 16:15:12.688811       629
2015-04-29 16:15:13.975050       829
2015-04-29 16:19:10.132692       209
2015-04-29 16:19:12.551835       599
2015-04-29 16:29:14.309592       889
2015-04-29 16:37:13.759633       779
2015-04-29 16:45:10.409904       259
2015-04-29 16:53:09.252755999     49
2015-04-29 16:59:09.398863        89
2015-04-29 17:17:11.879393       489
2015-04-29 17:19:11.738710       459
2015-04-29 17:39:14.174500       849
2015-04-29 17:43:09.832230       139
2015-04-29 18:05:13.977414       799
2015-04-29 18:39:14.808507       929
2015-04-29 18:53:09.611510        99
2015-04-29 18:53:11.228877       349
2015-04-29 19:03:09.612759        99
2015-04-29 19:23:09.425629        59
2015-04-29 19:23:12.258533       529
Name: hpies_eindex, dtype: int64

These seem to line up as expected.

We've already identified this data as problematic for the parser:

`#3__DE 69 396672 397189 397303 397438 396820 397744\\r*23f7`

We'll need to check why this record is failing:

`#3__DE 49 396688 397234 397354 397451 396786 397788*8ebd`

In [16]:
def crc3kerm(buf):
    """
    Compute the Kermit checksum on @a buf
    """
    crcta = [0, 4225, 8450, 12675, 16900, 21125, 25350, 29575,
             33800, 38025, 42250, 46475, 50700, 54925, 59150, 63375]
    crctb = [0, 4489, 8978, 12955, 17956, 22445, 25910, 29887,
             35912, 40385, 44890, 48851, 51820, 56293, 59774, 63735]
    crc = 0
    for i in range(0, len(buf)):
        c = crc ^ ord(buf[i])
        hi4 = (c & 240) >> 4
        lo4 = c & 15
        crc = (crc >> 8) ^ (crcta[hi4] ^ crctb[lo4])
    return crc


def chksumnmea(s):
    xor = 0
    for i in range(0, len(s)):
        xor ^= ord(s[i])
    return xor

def good_checksum(data):
    response, crc_str = raw.split('*')
    crc = int(crc_str, 16)
    calc_crc = crc3kerm(response)
    return crc == calc_crc

raw = jump_df.iloc[5].values[0]
good_checksum(raw)

True

CRC is good... so why did this fail?

Do we have any checksum failures, maybe my indexing is wrong...

In [17]:
[each[0] for each in raw_hef_df.values if not good_checksum(each[0])]

[]

Nope, all checksums pass...